### PR TITLE
forgejo: 1.18.0-rc1-1 -> 1.18.2-0

### PR DESCRIPTION
**Pull request by Izorkin**
###### Description of changes
Update forgejo to version 1.18.2-0

cc @SuperSandro2000
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by urandom2 — APPROVED** [review]

**Review by cyplo — APPROVED** [review]

**Comment by AmandaCameron** [comment]
I don't think this (or the initial patch) are embedding the correct version into the go binary, but instead embedding the version of the gitea binary. The `preBuild` step from [The original gitea drv](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/applications/version-management/gitea/default.nix#L37) probably needs to be copied in verbatium

**Comment by Izorkin** [comment]
> I don't think this (or the initial patch) are embedding the correct version into the go binary, but instead embedding the version of the gitea binary. The `preBuild` step from [The original gitea drv](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/applications/version-management/gitea/default.nix#L37) probably needs to be copied in verbatium

Added `preBuild` step.

File: pkgs/applications/version-management/forgejo/default.nix [hunk 1]
```diff
@@ -1,16 +1,32 @@
...
+  preBuild =
+    let
+      tags = lib.optional pamSupport "pam"
+        ++ lib.optional sqliteSupport "sqlite sqlite_unlock_notify";
+      tagsString = lib.concatStringsSep " " tags;
+    in
+    ''
+      export buildFlagsArray=(
+        -tags="${tagsString}"
+        -ldflags='-X "main.Version=${version}" -X "main.Tags=${tagsString}"'
```
Comment (urandom2) [inline pkgs/applications/version-management/forgejo/default.nix:26]: `ldflags` are canonically provided via attr: (meaning you would need to move your let to the top level)
```nix
ldflags = [
  "-X "main.Version=${version}"
  "-X main.Tags=${tagsString}"
];
```

File: pkgs/applications/version-management/forgejo/default.nix [hunk 2]
```diff
@@ -1,16 +1,32 @@
-{ fetchurl, gitea, lib }:
+{ lib, fetchurl, gitea
+, pamSupport ? true
+, sqliteSupport ? true
+}:
```
Comment (urandom2) [inline pkgs/applications/version-management/forgejo/default.nix:4]:
```suggestion
{ lib
, fetchurl
, gitea
, pamSupport ? true
, sqliteSupport ? true
}:
```

File: pkgs/applications/version-management/forgejo/default.nix [hunk 3]
```diff
@@ -1,16 +1,32 @@
...
+      tags = lib.optional pamSupport "pam"
+        ++ lib.optional sqliteSupport "sqlite sqlite_unlock_notify";
+      tagsString = lib.concatStringsSep " " tags;
+    in
+    ''
+      export buildFlagsArray=(
+        -tags="${tagsString}"
+        -ldflags='-X "main.Version=${version}" -X "main.Tags=${tagsString}"'
+      )
+    '';
```
Comment (urandom2) [inline pkgs/applications/version-management/forgejo/default.nix:28]: `ldflags` and `tags` are canonically provided via attr:
```suggestion
ldflags = [
  "-X main.Version=${version}"
  "-X main.Tags=${lib.concatStringsSep " " tags}"
];

tags = lib.optional pamSupport [ "pam" ] ++ lib.optional sqliteSupport [ "sqlite" "sqlite_unlock_notify" ];
```

Comment (Izorkin) [reply pkgs/applications/version-management/forgejo/default.nix:28]: Error build:
```
error: cannot coerce a list to a string

           20|     "-X main.Version=${version}"
           21|     "-X main.Tags=${lib.concatStringsSep " " tags}"
             |                     ^
           22|   ];
(use '--show-trace' to show detailed location information)

```

File: pkgs/applications/version-management/forgejo/default.nix [hunk 4]
```diff
@@ -1,16 +1,32 @@
...
   pname = "forgejo";
-  version = "1.18.0-rc1-1";
+  version = "1.18.0-1";
 
   src = fetchurl {
     name = "${pname}-src-${version}.tar.gz";
     # see https://codeberg.org/forgejo/forgejo/releases
-    url = "https://codeberg.org/attachments/976c426a-3e04-49ff-9762-47fab50624a3";
-    hash = "sha256-kreBMHlMVB1UeG67zMbszGrgjaROateCRswH7GrKnEw=";
+    url = "https://codeberg.org/attachments/c829784c-3b85-4996-9dc6-09e12e40a93a";
```
Comment (SuperSandro2000) [inline pkgs/applications/version-management/forgejo/default.nix:13]: I hope they fix this URL in their fork.

File: pkgs/applications/version-management/forgejo/default.nix (hunk 3 above)
Comment (SuperSandro2000) [reply pkgs/applications/version-management/forgejo/default.nix:28]: You want to use lib.optional**s** with lists

**Comment by Izorkin** [comment]
Updated PR.

**Comment by priegger** [comment]
Result of `nixpkgs-review pr 208623` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>forgejo</li>
  </ul>
</details>

**Comment by Izorkin** [comment]
Updated PR. The previous build variant no longer works.
```
forgejo> unpacking sources
forgejo> unpacking source archive /nix/store/9gsfc4nnb1hyrbmrffcdsrx81492idj2-forgejo-src-1.18.1-0.tar.gz
forgejo> source root is forgejo-src-1.18.1-0
forgejo> setting SOURCE_DATE_EPOCH to timestamp 1674007332 of file forgejo-src-1.18.1-0/vendor/xorm.io/xorm/tags/tag.go
forgejo> patching sources
forgejo> applying patch /nix/store/d8z8lzd6kqi4m6qv86nlsw0ylzvd5l6g-static-root-path.patch
forgejo> patching file modules/setting/setting.go
forgejo> Hunk #1 succeeded at 809 (offset 142 lines).
forgejo> configuring
forgejo> building
forgejo> Building subPackage code.gitea.io/gitea
forgejo> # code.gitea.io/gitea
forgejo> usage: link [options] main.o
forgejo>   -B note
forgejo>        add an ELF NT_GNU_BUILD_ID note when using ELF
forgejo>   -E entry
forgejo>        set entry symbol name
forgejo>   -H type
forgejo>        set header type
forgejo>   -I linker
forgejo>        use linker as ELF dynamic linker

```

File: pkgs/applications/version-management/forgejo/default.nix (hunk 3 above)
Comment (SuperSandro2000) [reply pkgs/applications/version-management/forgejo/default.nix:49]: Did you try to use the correct type when concatenating?

**Review by SuperSandro2000 — COMMENTED** [review]
LGTM except the tags/ldflags thread

File: pkgs/applications/version-management/forgejo/default.nix (hunk 3 above)
Comment (Izorkin) [reply pkgs/applications/version-management/forgejo/default.nix:49]: Haven't tried. Any possible examples?

Comment (Izorkin) [reply pkgs/applications/version-management/forgejo/default.nix:49]: This variant not working:
```diff
diff --git a/pkgs/applications/version-management/forgejo/default.nix b/pkgs/applications/version-management/forgejo/default.nix
index 7ba0ce7d438..c92ece7353b 100644
--- a/pkgs/applications/version-management/forgejo/default.nix
+++ b/pkgs/applications/version-management/forgejo/default.nix
@@ -12,6 +12,12 @@
 , sqliteSupport ? true
 }:

+let
+  tags = lib.optional pamSupport "pam"
+    ++ lib.optional sqliteSupport "sqlite sqlite_unlock_notify";
+  tagsString = lib.concatStringsSep " " tags;
+in
+
 buildGoPackage rec {
   pname = "forgejo";
   version = "1.18.0-1";
@@ -35,18 +41,10 @@ buildGoPackage rec {

   buildInputs = lib.optional pamSupport pam;

-  preBuild =
-    let
-      tags = lib.optional pamSupport "pam"
-        ++ lib.optional sqliteSupport "sqlite sqlite_unlock_notify";
-      tagsString = lib.concatStringsSep " " tags;
-    in
-    ''
-      export buildFlagsArray=(
-        -tags="${tagsString}"
-        -ldflags='-X "main.Version=${version}" -X "main.Tags=${tagsString}"'
-      )
-    '';
+  ldflags = [
+    "-X main.Version=${version}"
+    "-X main.Tags=${tagsString}"
+  ];

   outputs = [ "out" "data" ];

```

**Comment by Izorkin** [comment]
> LGTM except the tags/ldflags thread

Updated.

File: pkgs/applications/version-management/forgejo/default.nix [hunk 5]
```diff
@@ -1,25 +1,72 @@
...
+  postPatch = ''
+    substituteInPlace modules/setting/setting.go --subst-var data
   '';
 
+  preBuild = ''
+    export buildFlagsArray=(
+      -tags="${tagsString}"
+      -ldflags='-X "main.Version=${version}" -X "main.Tags=${tagsString}"'
+    )
+  '';
```
Comment (SuperSandro2000) [inline pkgs/applications/version-management/forgejo/default.nix:51]:
```suggestion
  tags = lib.optional pamSupport "pam"
    ++ lib.optional sqliteSupport "sqlite sqlite_unlock_notify";
  
  ldflags = [
    "-X main.Version=${version}"
    "-X main.Tags=${lib.concatStringsSep " " tags}"
  ];
```

File: pkgs/applications/version-management/forgejo/default.nix [hunk 6]
```diff
@@ -1,25 +1,72 @@
...
+, sqliteSupport ? true
+}:
 
-gitea.overrideAttrs (old: rec {
+let
+  tags = lib.optional pamSupport "pam"
+    ++ lib.optional sqliteSupport "sqlite sqlite_unlock_notify";
+  tagsString = lib.concatStringsSep " " tags;
+in
+
```
Comment (SuperSandro2000) [inline pkgs/applications/version-management/forgejo/default.nix:20]:
```suggestion
```

File: pkgs/applications/version-management/forgejo/default.nix [hunk 7]
```diff
@@ -1,25 +1,72 @@
...
+buildGoPackage rec {
   pname = "forgejo";
-  version = "1.18.0-rc1-1";
+  version = "1.18.0-1";
```
Comment (IndeedNotJames) [inline pkgs/applications/version-management/forgejo/default.nix:23]:
```suggestion
  version = "1.18.2-0";
```

File: pkgs/applications/version-management/forgejo/default.nix [hunk 8]
```diff
@@ -1,25 +1,72 @@
...
 
   src = fetchurl {
     name = "${pname}-src-${version}.tar.gz";
     # see https://codeberg.org/forgejo/forgejo/releases
-    url = "https://codeberg.org/attachments/976c426a-3e04-49ff-9762-47fab50624a3";
-    hash = "sha256-kreBMHlMVB1UeG67zMbszGrgjaROateCRswH7GrKnEw=";
+    url = "https://codeberg.org/attachments/c829784c-3b85-4996-9dc6-09e12e40a93a";
+    hash = "sha256-42bR18T5ATVyhPejeHpOTkeP6VwYu5HFfei8oMjRJy8=";
```
Comment (IndeedNotJames) [inline pkgs/applications/version-management/forgejo/default.nix:29]:
```suggestion
    url = "https://codeberg.org/attachments/5d59ec04-9f29-4b32-a1ef-bec5c3132e26";
    hash = "sha256-RLShwdx8geyFr1Jk5qDVbsEt2hCjdrwX0lNHea7P+pk=";
```
for [v1.18.2-0](https://codeberg.org/forgejo/forgejo/releases/tag/v1.18.2-0)

**Review by emilylange — CHANGES_REQUESTED** [review]
[v1.18.2-0](https://codeberg.org/forgejo/forgejo/releases/tag/v1.18.2-0) got released :)

Release notes: https://codeberg.org/forgejo/forgejo/src/commit/9d5805245458931cf490f9360693430d65156386/RELEASE-NOTES.md#1-18-2-0

File: pkgs/applications/version-management/forgejo/default.nix (hunk 5 above)
Comment (Izorkin) [reply pkgs/applications/version-management/forgejo/default.nix:51]: Error build:
```
Building subPackage code.gitea.io/gitea
cannot find package "sqlite_unlock_notify" in any of:
        /nix/store/x4p5yg47ai209jin1phhj4fc2iw3b5ix-go-1.19.4/share/go/src/sqlite_unlock_notify (from $GOROOT)
        /build/go/src/sqlite_unlock_notify (from $GOPATH)
cannot find package "-ldflags=-X main.Version=1.18.2-0 -X main.Tags=pam sqlite sqlite_unlock_notify" in any of:
        /nix/store/x4p5yg47ai209jin1phhj4fc2iw3b5ix-go-1.19.4/share/go/src/-ldflags=-X main.Version=1.18.2-0 -X main.Tags=pam sqlite sqlite_unlock_notify (from $GOROOT)
        /build/go/src/-ldflags=-X main.Version=1.18.2-0 -X main.Tags=pam sqlite sqlite_unlock_notify (from $GOPATH)
cannot find package "-p" in any of:
        /nix/store/x4p5yg47ai209jin1phhj4fc2iw3b5ix-go-1.19.4/share/go/src/-p (from $GOROOT)
        /build/go/src/-p (from $GOPATH)
cannot find package "32" in any of:
        /nix/store/x4p5yg47ai209jin1phhj4fc2iw3b5ix-go-1.19.4/share/go/src/32 (from $GOROOT)
        /build/go/src/32 (from $GOPATH)

```

Comment (IndeedNotJames) [reply pkgs/applications/version-management/forgejo/default.nix:51]: Use
```suggestion
  tags = lib.optional pamSupport "pam"
    ++ lib.optionals sqliteSupport [ "sqlite" "sqlite_unlock_notify" ];

  ldflags = [
    "-X main.Version=${version}"
    "-X 'main.Tags=${lib.concatStringsSep " " tags}'"
  ];
```
instead.
Note the `lib.optionals` and single quotes `'` in `"-X 'main.Tags=${lib.concatStringsSep " " tags}'"`

```
❯ nix-build -A forgejo
/nix/store/k818d0m46hc380cclhdrp0sfkyi414hc-forgejo-1.18.2-0
❯ ./result/bin/gitea --version
Gitea version 1.18.2-0 built with go1.19.4 : pam, sqlite, sqlite_unlock_notify
```

Comment (Izorkin) [reply pkgs/applications/version-management/forgejo/default.nix]: Thanks!

**Review by emilylange — APPROVED** [review]

File: pkgs/applications/version-management/forgejo/default.nix [hunk 9]
```diff
@@ -1,25 +1,66 @@
...
+, pam
+, pamSupport ? true
+, sqliteSupport ? true
+}:
 
-gitea.overrideAttrs (old: rec {
+buildGoPackage rec {
   pname = "forgejo";
-  version = "1.18.0-rc1-1";
+  version = "1.18.2-0";
```
Comment (SuperSandro2000) [inline pkgs/applications/version-management/forgejo/default.nix:17]: If the suffix is always -0 we should drop it in the future.

Comment (Izorkin) [reply pkgs/applications/version-management/forgejo/default.nix:17]: Hm... This is correct version:
```
./result/bin/gitea --version
Gitea version 1.18.2-0 built with go1.19.4 : pam, sqlite, sqlite_unlock_notify
```

Comment (SuperSandro2000) [reply pkgs/applications/version-management/forgejo/default.nix:17]: We are hardcoding the version in line 43, so I am not surprised that it outputs this.

Comment (SuperSandro2000) [reply pkgs/applications/version-management/forgejo/default.nix:17]: The dash one is sometimes used, so we keep it https://codeberg.org/forgejo/-/packages/container/forgejo/1.18.2-1

Comment (IndeedNotJames) [reply pkgs/applications/version-management/forgejo/default.nix:17]: Yeah, just for future reference, the versioning is explained in
https://codeberg.org/forgejo/forgejo/src/branch/forgejo/CONTRIBUTING/RELEASE.md#release-numbering [[permalink](https://codeberg.org/forgejo/forgejo/src/commit/0db7131d5bd07a2c8b7874457998551fd601c6c4/CONTRIBUTING/RELEASE.md#release-numbering)]

**Pull request merged**